docs/src/AdvancedUsageGuide.md

```diff
@@ -0,0 +1,488 @@
+│   %3  = Base.broadcasted(ITensors.ind, ivs)::Base.Broadcast.Broadcasted{Base.Broadcast.Style{Tuple},Nothing,typeof(ind),Tuple{Tuple{Pair{Index{Int64},Int64},Pair{Index{Int64},Int64}}}}
+│   %4  = Base.materialize(%3)::Tuple{Index{Int64},Index{Int64}}
+│         (p = (%1)(%2, %4))
+│   %6  = NDTensors.permute::Core.Compiler.Const(NDTensors.permute, false)
+│   %7  = Base.broadcasted(ITensors.val, ivs)::Base.Broadcast.Broadcasted{Base.Broadcast.Style{Tuple},Nothing,typeof(val),Tuple{Tuple{Pair{Index{Int64},Int64},Pair{Index{Int64},Int64}}}}
+│   %8  = Base.materialize(%7)::Tuple{Int64,Int64}
+│         (vals = (%6)(%8, p))
+│   %10 = Core.tuple(T)::Tuple{ITensor{2}}
+│   %11 = Core._apply_iterate(Base.iterate, Base.getindex, %10, vals)::Any
+└──       return %11
+```
+Uh oh, that doesn't look good! Julia can't know ahead of time, based on 
+the inputs, what the type of the output is (though at runtime, the 
+output has a concrete type).
+
+So why is it designed this way? The main reason is to allow more 
+generic code. This allows us to have code like:
+```julia
+A = randomITensor(i', i)
+A .*= 2+1im
+```
+Here, the type of the storage of A is changed in-place. More 
+generally, this allows ITensors to have more generic in-place 
+functionality, so you can write code where you don't know what the
+storage is until runtime.
+
+This can lead to certain types of code having perfomance problems, 
+for example looping through ITensors can be slow:
+```julia
+A = randomITensor(i', i)
+for n in 1:dim(A)
+  A[n] = 2 * A[n]
+end
+```
+However, this is fast:
+```julia
+A .*= 2
+```
+How does this work? It relies on a "function barrier" technique. 
+Julia compiles functions "just-in-time", so that calls an inner 
+function written in terms of a type-stable Tensor type from the 
+package NDTensors. That is the function ultimately being called. 
+The main overhead is that Julia has to determine which function 
+to call at runtime.
+
+You can look at a simple example to see how this works.
+```julia
+struct MyMatrix
+  data
+end
+
+import Base: *
+
+# This is our function barrier
+A::MyMatrix * B::MyMatrix = A.data * B.data
+
+for d in 5:5:40
+  a = randn(d,d)
+  b = randn(d,d)
+
+  A = MyMatrix(a)
+  B = MyMatrix(b)
+
+  @btime $a*$b
+  @btime $A*$B
+end
+```
+However,
+```julia
+d = 100
+
+a = randn(d,d)
+
+A = MyMatrix(a)
+
+function Base.sum(A::MyMatrix)
+  t = 0.0
+  for i in 1:length(A.data)
+    t += A.data[i]
+  end
+  return A
+end
+```
+
+One can make code type stable by putting explicit type declerations, 
+if they know the type beforehand. For example:
+```julia
+function Base.sum(A::MyMatrix)
+  t::Float64 = 0.0
+  for i in 1:length(A.data)
+    t += A.data[i]
+  end
+  return A
+end
+```
+
+Therefore, users should keep this in mind when they are writing 
+ITensors.jl code, and we warn that explicitly looping over large 
+ITensor by individual elements should be avoiding in performance 
```

Comment:
Maybe rephrase to "should be done with caution", since there may be ways to make it fast with type declarations.

Reply:
That sounds good. What about also pointing to some alternatives to looping over ITensor elements, such as using broadcasted calls?

Reply:
Yes of course, that is also discussed in the next section but can be pointed out here.